HPCC-20426 cryptohelper into jlib dll, use openssl routines where available, refactor digisign

jakesmith opened the pull request:
<!-- Thank you for submitting a pull request to the HPCC project

 PLEASE READ the following before proceeding.

 This project only accepts pull requests related to open JIRA issues.
 If suggesting a new feature or change, please discuss it in a JIRA issue first.
 If fixing a bug, there should be an issue describing it with steps to reproduce.
 The title line of the pull request (and of each commit within it) should refer to the
 associated issue using the format:

 HPCC-nnnnn Short description of issue

 This will allow the Jira ticket to be automatically updated to refer to this pull request,
 and will ensure that the automatically-generated changelog is properly formatted.
 Where a pull request contains a single commit the pull request title will be set automatically,
 assuming that the commit has followed the proper guidelines.

 Please go over all the following points, and put an `x` in all the boxes that apply. You may find
 it easier to press the 'Create' button first then click on the checkboxes to edit the comment.
-->

## Type of change:
- [ ] This change is a bug fix (non-breaking change which fixes an issue).
- [ ] This change is a new feature (non-breaking change which adds functionality).
- [x] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [ ] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [ ] The commit message title makes sense in a changelog, by itself.
  - [ ] The commit is signed.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [x] I have read the CONTRIBUTORS document.
- [x] The change has been fully tested:
  - [ ] I have added tests to cover my changes.
  - [ ] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [x] I have given due consideration to all of the following potential concerns:
  - [ ] Scalability
  - [ ] Performance
  - [ ] Security
  - [ ] Thread-safety
  - [ ] Premature optimization
  - [ ] Existing deployed queries will not be broken
  - [ ] This change fixes the problem, not just the symptom
  - [ ] The target branch of this pull request is appropriate for such a change.
- [x] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

## Testing:
<!-- Please describe how this change has been tested.-->

<!-- Thank you for taking the time to submit this pull request and to answer all of the above-->

jakesmith commented:
@ghalliday - based off hpcc-20383. Please take a look at last 2 commits.

HPCCSmoketest commented:
Automated Smoketest: :white_check_mark:
OS: centos 7.4.1708 (Linux 3.10.0-327.28.3.el7.x86_64)
Sha: 9eca57e0358f9cb73dc40be475af3ea504c74d88
Build: success
Install hpccsystems-platform-community_7.0.0-rc1.el7.x86_64.rpm
HPCC Start: OK

Unit tests result:

| Test | total | passed | failed | errors | timeout | 
|---|---|---|---|---|---|
|unittest |94 |94 |0 |0 |0 |
|wutoolTest(Dali) |19 |19 |0 |0 |0 |
|wutoolTest(Cassandra) |19 |19 |0 |0 |0 |



Regression test result:

| phase | total | pass | fail | 
|---|---|---|---|
|setup (hthor) |11 |11 |0 |
|setup (thor) |11 |11 |0 |
|setup (roxie) |11 |11 |0 |
|test (hthor) |805 |805 |0 |
|test (thor) |729 |729 |0 |
|test (roxie) |878 |878 |0 |


HPCC Stop: OK
HPCC Uninstall: OK
Time stats:

| Prep time | Build time | Package time | Install time | Start time | Test time | Stop time | Summary | 
|---|---|---|---|---|---|---|---|
|   20 sec (00:00:20) |  176 sec (00:02:56) |   93 sec (00:01:33) |   21 sec (00:00:21) |   43 sec (00:00:43) | 1373 sec (00:22:53) |   19 sec (00:00:19) | 1745 sec (00:29:05) |

HPCCSmoketest commented:
Automated Smoketest: :x:
OS: centos 7.4.1708 (Linux 3.10.0-327.28.3.el7.x86_64)
Sha: 2fe50d51f954cfd2f0f4d3af62cd87942575ce6f
Build: success
Install hpccsystems-platform-community_7.0.0-rc1.el7.x86_64.rpm
HPCC Start: Fail
HPCC start:
Starting mydafilesrv ...     [   OK    ]  
Starting mydali ...          [   OK    ]  
Starting mydfuserver ...     [   OK    ]  
Starting myeclagent ...      [   OK    ]  
Starting myeclccserver ...   [   OK    ]  
Starting myeclscheduler ...  [   OK    ]  
Starting myesp ...           [   OK    ]  
Starting myroxie ...         [ TIMEOUT ] 
Starting mysasha ...         [   OK    ]  
Starting mythor ...          [   OK    ]  
 
HPCC status:
mydafilesrv     ( pid     9387 ) is running ...
mydali          ( pid     9710 ) is running ...
mydfuserver     ( pid    10059 ) is running ...
myeclagent      ( pid    10408 ) is running ...
myeclccserver   ( pid    10741 ) is running ...
myeclscheduler  ( pid    11088 ) is running ...
myesp           ( pid    11427 ) is running ...
myroxie         is stopped,
mysasha         ( pid    13043 ) is running ...
mythor          ( pid    13627 ) is running with 1 slave process(es) ...
myroxie         is stopped,
HPCC Uninstall: OK
Time stats:

| Prep time | Build time | Package time | Install time | Start time | Test time | Stop time | Summary | 
|---|---|---|---|---|---|---|---|
|   12 sec (00:00:12) |  173 sec (00:02:53) |   93 sec (00:01:33) |   22 sec (00:00:22) |  164 sec (00:02:44) |    0 sec (00:00:00) |    0 sec (00:00:00) |  464 sec (00:07:44) |

HPCCSmoketest commented:
Automated Smoketest: :white_check_mark:
OS: centos 7.4.1708 (Linux 3.10.0-327.28.3.el7.x86_64)
Sha: edc06062e6f32fda735645a1579e913a54255042
Build: success
Install hpccsystems-platform-community_7.0.0-rc1.el7.x86_64.rpm
HPCC Start: OK

Unit tests result:

| Test | total | passed | failed | errors | timeout | 
|---|---|---|---|---|---|
|unittest |94 |94 |0 |0 |0 |
|wutoolTest(Dali) |19 |19 |0 |0 |0 |
|wutoolTest(Cassandra) |19 |19 |0 |0 |0 |



Regression test result:

| phase | total | pass | fail | 
|---|---|---|---|
|setup (hthor) |11 |11 |0 |
|setup (thor) |11 |11 |0 |
|setup (roxie) |11 |11 |0 |
|test (hthor) |805 |805 |0 |
|test (thor) |729 |729 |0 |
|test (roxie) |878 |878 |0 |


HPCC Stop: OK
HPCC Uninstall: OK
Time stats:

| Prep time | Build time | Package time | Install time | Start time | Test time | Stop time | Summary | 
|---|---|---|---|---|---|---|---|
|   10 sec (00:00:10) |  178 sec (00:02:58) |   93 sec (00:01:33) |   21 sec (00:00:21) |   43 sec (00:00:43) | 1358 sec (00:22:38) |   18 sec (00:00:18) | 1721 sec (00:28:41) |

HPCCSmoketest commented:
Automated Smoketest: :white_check_mark:
OS: centos 7.4.1708 (Linux 3.10.0-327.28.3.el7.x86_64)
Sha: 8bbfc9d4a632ba45d55a1d4b13cea734c549e675
Build: success
Install hpccsystems-platform-community_7.0.0-rc1.el7.x86_64.rpm
HPCC Start: OK

Unit tests result:

| Test | total | passed | failed | errors | timeout | 
|---|---|---|---|---|---|
|unittest |94 |94 |0 |0 |0 |
|wutoolTest(Dali) |19 |19 |0 |0 |0 |
|wutoolTest(Cassandra) |19 |19 |0 |0 |0 |



Regression test result:

| phase | total | pass | fail | 
|---|---|---|---|
|setup (hthor) |11 |11 |0 |
|setup (thor) |11 |11 |0 |
|setup (roxie) |11 |11 |0 |
|test (hthor) |805 |805 |0 |
|test (thor) |729 |729 |0 |
|test (roxie) |878 |878 |0 |


HPCC Stop: OK
HPCC Uninstall: OK
Time stats:

| Prep time | Build time | Package time | Install time | Start time | Test time | Stop time | Summary | 
|---|---|---|---|---|---|---|---|
|   28 sec (00:00:28) |  173 sec (00:02:53) |   94 sec (00:01:34) |   21 sec (00:00:21) |   43 sec (00:00:43) | 1337 sec (00:22:17) |   19 sec (00:00:19) | 1715 sec (00:28:35) |

jakesmith committed:
HPCC-20426 Refactor crypto helper and compile into jlib.

+ Modify digisign to use CLoadedKey (to avoid reloading keys)
+ Common up openSSL initialization, e.g. locking callbacks
+ Merge compilation into jlib, to avoid another dependency.

Signed-off-by: Jake Smith <jake.smith@lexisnexisrisk.com>

jakesmith commented:
@ghalliday - I managed to build under windows and have resolved the build issue.

This PR also contains some refactoring of the digisign, to avoid some of threading issues and be more efficient + common up the setting up of ssl callbacks for locking etc.
So this PR also resolves - HPCC-20345

Please review and double check windows compilation.

HPCCSmoketest commented:
Automated Smoketest: :white_check_mark:
OS: centos 7.4.1708 (Linux 3.10.0-327.28.3.el7.x86_64)
Sha: 262e992951076dcba4e9a12765d6fbcf27093e6c
Build: success
Install hpccsystems-platform-community_7.0.0-rc1.el7.x86_64.rpm
HPCC Start: OK

Unit tests result:

| Test | total | passed | failed | errors | timeout | 
|---|---|---|---|---|---|
|unittest |89 |89 |0 |0 |0 |
|wutoolTest(Dali) |19 |19 |0 |0 |0 |
|wutoolTest(Cassandra) |19 |19 |0 |0 |0 |



Regression test result:

| phase | total | pass | fail | 
|---|---|---|---|
|setup (hthor) |11 |11 |0 |
|setup (thor) |11 |11 |0 |
|setup (roxie) |11 |11 |0 |
|test (hthor) |805 |805 |0 |
|test (thor) |729 |729 |0 |
|test (roxie) |878 |878 |0 |


HPCC Stop: OK
HPCC Uninstall: OK
Time stats:

| Prep time | Build time | Package time | Install time | Start time | Test time | Stop time | Summary | 
|---|---|---|---|---|---|---|---|
|   30 sec (00:00:30) |  173 sec (00:02:53) |   93 sec (00:01:33) |   20 sec (00:00:20) |   42 sec (00:00:42) | 1333 sec (00:22:13) |   19 sec (00:00:19) | 1710 sec (00:28:30) |

ghalliday reviewed (commented):
@jakesmith a few comments.  Mostly reinstating tests that have been removed, and other minor comments.

jakesmith commented on testing/unittests/cryptotests.cpp:
```diff
@@ -85,11 +85,11 @@ class CryptoUnitTest : public CppUnit::TestFixture
-        CPPUNIT_TEST(aesEncryptDecryptTests);
-        CPPUNIT_TEST(aesWithRsaEncryptedKey);
-        CPPUNIT_TEST(aesParallelTest);
+//        CPPUNIT_TEST(pkeEncryptDecryptTest);
```
whoops yes

jakesmith commented on testing/unittests/cryptotests.cpp:
```diff
@@ -201,90 +208,90 @@ class CryptoUnitTest : public CppUnit::TestFixture
-
+            Owned<IDigitalSignatureManager> dsm(createDigitalSignatureManagerInstanceFromKeys(pubKey, privKey, nullptr));
 
+/*
```
will do

jakesmith commented on system/jlib/jencrypt.cpp:
```diff
@@ -1798,8 +1804,29 @@ MemoryBuffer &aesDecrypt(const void *key, size_t keylen, const void *input, size
+MemoryBuffer &aesEncrypt(const void *key, size_t keylen, const void *input, size_t inlen, MemoryBuffer &output)
+{
+#if defined(_USE_OPENSSL) && !defined(_WIN32)
+	cryptohelper::aesEncrypt(output, inlen, input, keylen, (const char *)key);
```
it's a tab, not sure how that got there, will fix.

jakesmith commented on system/security/cryptohelper/digisign.cpp:
```diff
@@ -277,79 +194,101 @@ static void createDigitalSignatureManagerInstance(IDigitalSignatureManager * * p
         }
-
-        if (!isEmptyString(_privKey))
+        catch (IException * e)
```
I don't really like the IDigitalSignatureManager conflating the two. I introduced separate digiSign and digiVerify standalone methods which is what HPCC-20340 will use (digiSign in esp and digiVerify in dafilesrv).

I will change it report the error but allow it to continue.

jakesmith commented on system/security/cryptohelper/digisign.hpp:
```diff
@@ -27,31 +27,41 @@
+    virtual bool isDigiSignerConfigured() const = 0;
+    virtual bool isDigiVerifierConfigured() const = 0;
+    virtual bool digiSign(StringBuffer & b64Signature, const char * text) const = 0;//signs, using private key
+    virtual bool digiVerify(StringBuffer & b64Signature, const char * text) const = 0;//verifies, using public key
```
yeah I agree, I will change.

jakesmith commented on system/security/cryptohelper/cryptocommon.cpp:
```diff
@@ -17,21 +17,51 @@
+    numCryptoLocks = CRYPTO_num_locks();
+    for (unsigned i=0; i<numCryptoLocks; i++)
+        cryptoLocks.push_back(std::unique_ptr<CriticalSection>(new CriticalSection));
+    CRYPTO_set_locking_callback(locking_function);
```
have changed.

jakesmith commented on system/security/cryptohelper/digisign.hpp:
```diff
@@ -27,31 +27,41 @@
+bool digiSign(StringBuffer &b64Signature, const char *text, const CLoadedKey &signingKey);
+
+//Verify the given text was used to create the given digital signature
+bool digiVerify(StringBuffer &b64Signature, const char *text, const CLoadedKey &verifyingKey);
```
agreed, will change

jakesmith commented:
@ghalliday - please see review changes.

ghalliday commented:
@jakesmith I think you haven't reversed the parameters on the two calls to digiVerify in daldap.cpp

HPCCSmoketest commented:
Automated Smoketest: :x:
OS: centos 7.4.1708 (Linux 3.10.0-327.28.3.el7.x86_64)
Sha: 9317b72a6bdb08966c85cf63572e9afe6dcfee6f
Build: success
Install hpccsystems-platform-community_7.0.0-rc1.el7.x86_64.rpm
HPCC Start: OK

Unit tests result:

| Test | total | passed | failed | errors | timeout | 
|---|---|---|---|---|---|
|unittest |94 |94 |0 |0 |0 |
|wutoolTest(Dali) |19 |19 |0 |0 |0 |
|wutoolTest(Cassandra) |19 |19 |0 |0 |0 |



Regression test result:

| phase | total | pass | fail | 
|---|---|---|---|
|setup (hthor) |11 |3 |8 |
|setup (thor) |11 |3 |8 |
|setup (roxie) |11 |3 |8 |
|test (hthor) |805 |438 |367 |
|test (thor) |729 |398 |331 |
|test (roxie) |878 |476 |402 |


[Errors:](https://gist.github.com/42b1dd2f67a9305062a6b2e5d7937e71)
- setup
  - hthor
    - 3\. [setup](https://gist.github.com/HPCCSmoketest/42b1dd2f67a9305062a6b2e5d7937e71#file-setup-w20180906-163739-9-log), 4\. [setup_fetch](https://gist.github.com/HPCCSmoketest/42b1dd2f67a9305062a6b2e5d7937e71#file-setup_fetch-w20180906-163739-10-log), 5\. [setupdict](https://gist.github.com/HPCCSmoketest/42b1dd2f67a9305062a6b2e5d7937e71#file-setupdict-w20180906-163739-4-log), 6\. [setuppersist](https://gist.github.com/HPCCSmoketest/42b1dd2f67a9305062a6b2e5d7937e71#file-setuppersist-w20180906-163739-1-log), 7\. [setupsearchindex](https://gist.github.com/HPCCSmoketest/42b1dd2f67a9305062a6b2e5d7937e71#file-setupsearchindex-w20180906-163739-3-log), 8\. [setupsq](https://gist.github.com/HPCCSmoketest/42b1dd2f67a9305062a6b2e5d7937e71#file-setupsq-w20180906-163739-7-log), 9\. [setupwordindex](https://gist.github.com/HPCCSmoketest/42b1dd2f67a9305062a6b2e5d7937e71#file-setupwordindex-w20180906-163739-6-log), 10\. [setupxml](https://gist.github.com/HPCCSmoketest/42b1dd2f67a9305062a6b2e5d7937e71#file-setupxml-w20180906-163739-12-log), 
  - thor
    - 3\. [setup](https://gist.github.com/HPCCSmoketest/42b1dd2f67a9305062a6b2e5d7937e71#file-setup-w20180906-163755-3-log), 4\. [setup_fetch](https://gist.github.com/HPCCSmoketest/42b1dd2f67a9305062a6b2e5d7937e71#file-setup_fetch-w20180906-163755-1-log), 5\. [setupdict](https://gist.github.com/HPCCSmoketest/42b1dd2f67a9305062a6b2e5d7937e71#file-setupdict-w20180906-163755-5-log), 6\. [setuppersist](https://gist.github.com/HPCCSmoketest/42b1dd2f67a9305062a6b2e5d7937e71#file-setuppersist-w20180906-163755-7-log), 7\. [setupsearchindex](https://gist.github.com/HPCCSmoketest/42b1dd2f67a9305062a6b2e5d7937e71#file-setupsearchindex-w20180906-163755-2-log), 8\. [setupsq](https://gist.github.com/HPCCSmoketest/42b1dd2f67a9305062a6b2e5d7937e71#file-setupsq-w20180906-163755-13-log), 9\. [setupwordindex](https://gist.github.com/HPCCSmoketest/42b1dd2f67a9305062a6b2e5d7937e71#file-setupwordindex-w20180906-163755-4-log), 10\. [setupxml](https://gist.github.com/HPCCSmoketest/42b1dd2f67a9305062a6b2e5d7937e71#file-setupxml-w20180906-163755-8-log), 
  - roxie
    - 3\. [setup](https://gist.github.com/HPCCSmoketest/42b1dd2f67a9305062a6b2e5d7937e71#file-setup-w20180906-163831-2-log), 4\. [setup_fetch](https://gist.github.com/HPCCSmoketest/42b1dd2f67a9305062a6b2e5d7937e71#file-setup_fetch-w20180906-163831-5-log), 5\. [setupdict](https://gist.github.com/HPCCSmoketest/42b1dd2f67a9305062a6b2e5d7937e71#file-setupdict-w20180906-163831-6-log), 6\. [setuppersist](https://gist.github.com/HPCCSmoketest/42b1dd2f67a9305062a6b2e5d7937e71#file-setuppersist-w20180906-163831-4-log), 7\. [setupsearchindex](https://gist.github.com/HPCCSmoketest/42b1dd2f67a9305062a6b2e5d7937e71#file-setupsearchindex-w20180906-163831-7-log), 8\. [setupsq](https://gist.github.com/HPCCSmoketest/42b1dd2f67a9305062a6b2e5d7937e71#file-setupsq-w20180906-163831-3-log), 9\. [setupwordindex](https://gist.github.
 ... ( comment is too long, 53 lines, 188721 bytes, truncated) ... 
Time stats:

| Prep time | Build time | Package time | Install time | Start time | Test time | Stop time | Summary | 
|---|---|---|---|---|---|---|---|
|   24 sec (00:00:24) |  173 sec (00:02:53) |   94 sec (00:01:34) |   19 sec (00:00:19) |   41 sec (00:00:41) | 1350 sec (00:22:30) |   19 sec (00:00:19) | 1720 sec (00:28:40) |

```

ghalliday commented:
Other than the comment above it looks ready to merge.  Please can you squash once you have addressed the verify calls.

Note the changes do add a noticeable improvement on the digiSign/digiVerify timings.
On my machine (only 1 run) Signing 1941 -> 1474  Verify 57->41, async signing improved even more.

jakesmith commented:
@ghalliday - please merge after smoketest passes.

HPCCSmoketest commented:
Automated Smoketest: :white_check_mark:
OS: centos 7.4.1708 (Linux 3.10.0-327.28.3.el7.x86_64)
Sha: a23c46f82494633290d5904b4d76a284cfc11a75
Build: success
Install hpccsystems-platform-community_7.0.0-rc1.el7.x86_64.rpm
HPCC Start: OK

Unit tests result:

| Test | total | passed | failed | errors | timeout | 
|---|---|---|---|---|---|
|unittest |94 |94 |0 |0 |0 |
|wutoolTest(Dali) |19 |19 |0 |0 |0 |
|wutoolTest(Cassandra) |19 |19 |0 |0 |0 |



Regression test result:

| phase | total | pass | fail | 
|---|---|---|---|
|setup (hthor) |11 |11 |0 |
|setup (thor) |11 |11 |0 |
|setup (roxie) |11 |11 |0 |
|test (hthor) |805 |805 |0 |
|test (thor) |729 |729 |0 |
|test (roxie) |878 |878 |0 |


HPCC Stop: OK
HPCC Uninstall: OK
Time stats:

| Prep time | Build time | Package time | Install time | Start time | Test time | Stop time | Summary | 
|---|---|---|---|---|---|---|---|
|   15 sec (00:00:15) |  174 sec (00:02:54) |   94 sec (00:01:34) |   21 sec (00:00:21) |   44 sec (00:00:44) | 1344 sec (00:22:24) |   17 sec (00:00:17) | 1709 sec (00:28:29) |

AttilaVamos commented:
@jakesmith The last commit: 895f504 won't be tested, because the PR is merged/closed.

jakesmith commented:
@AttilaVamos - that's fine - it was a WIN32 addition only, so smoketest wouldn't see any difference anyway.

AttilaVamos commented:
@jakesmith OK, thanks
